### ISSUE TITLE
Media query styles applied only to the first element in function

### ISSUE THREAD

**Issue by karpcs** [cssinjs/jss#1320]
__Expected behavior:__
media query should apply styles to all the component that uses it

__Describe the bug:__
only the first component gets the styles of media query

__Codesandbox link:__ 
https://codesandbox.io/s/react-jss-playground-bz42d

__Versions (please complete the following information):__
 - jss:10.1.1
 - Browser saw it on chrome, havent tested others
 - OS windows10:

![image](https://user-images.githubusercontent.com/43632837/77542171-6ec52480-6eae-11ea-8953-34fa6f6fa177.png)

**Comment by pranshuchittora** [comment]
@kof if this issue is still open then I would like to work on this :)

**Comment by kof** [comment]
@pranshuchittora sure

**Comment by pranshuchittora** [comment]
Thanks @kof  for the blazing fast reply :+1:

**Comment by pranshuchittora** [comment]
```diff
-  wrapper: () => ({
    padding: 40,
    background: theme.background,
    textAlign: "left",
    "@media (min-width: 1024px)": {
      backgroundColor: "red"
    }
-  }),
+ wrapper: {
    padding: 40,
    background: theme.background,
    textAlign: "left",
    "@media (min-width: 1024px)": {
      backgroundColor: "red"
    },
+  }, 
```

@karpcs  Converting the style object from an arrow function to plain object works.

![Screenshot from 2020-05-05 19-32-45](https://user-images.githubusercontent.com/32242596/81075290-c58b3880-8f07-11ea-883c-2c080beb5d4e.png)

@kof probably this is happening due to improper object referencing in arrow function approach. Is it still a bug, if no then I suggest closing this issue :)

**Comment by kof** [comment]
Feel free to dig deeper and figure out what the actual bug is. Yes it is only when rule function is used.

**Comment by kof** [comment]
Same as https://github.com/cssinjs/jss/issues/1327

**Comment by pranshuchittora** [comment]
Rule function is not transpiling ?
![Screenshot from 2020-05-05 20-25-04](https://user-images.githubusercontent.com/32242596/81080756-bb206d00-8f0e-11ea-8bff-c368301d85a7.png)

**Comment by kof** [comment]
Might be just the playground, also it's not on the latest version, feel free to update and check what it does.

**Comment by pranshuchittora** [comment]
I have updated the deps.
```json
    "jss": "^10.1.1",
    "jss-preset-default": "^10.1.1"
```

![Screenshot from 2020-05-05 20-47-02](https://user-images.githubusercontent.com/32242596/81084384-69c6ac80-8f13-11ea-8db6-07363a96394c.png)

Still not working, will take a look at the rule function parsing.
Is [this](https://github.com/cssinjs/jss/blob/d9cca31ad2d058aa0108fa6ec48fe872b0e7e886/packages/jss/src/Jss.js#L113) the correct function to look at ?

**Comment by pranshuchittora** [comment]
Hi @kof , Did some preliminary research and found out that the function is not being executed to get the returned object as shown below.

![Screenshot from 2020-05-06 03-22-26](https://user-images.githubusercontent.com/32242596/81119698-ef188400-8f48-11ea-9797-056ea94f2034.png)

As the function is not bein executed, therefore converting it toCss string returns `null`.

**Comment by karpcs** [comment]
I kept the version in playground at the version, that this bug appears, prior versions does not have this bug.

i noticed this when looping through array and passing different props to each and only the first element got the styles, so i just put a little demo here

**Comment by kof** [comment]
@pranshuchittora yes, you are still on the surface, sheet.update is supposed to call function rules, jss-plugin-rule-value-function plugin is responsible for implementing the function rule API ... somewhere between jss core and the plugin there is a bug, most likely in the plugin.

**Comment by kof** [comment]
https://github.com/cssinjs/repl/blob/gh-pages/src/index.js#L55 repl doesn't call .update, so on the repl function rules won't work.

The actual issue in your production code might be in react-jss

**Comment by pranshuchittora** [comment]
> @pranshuchittora yes, you are still on the surface, sheet.update is supposed to call function rules, jss-plugin-rule-value-function plugin is responsible for implementing the function rule API ... somewhere between jss core and the plugin there is a bug, most likely in the plugin.

Yeah, after call `update()` it worked. I have raised a PR regarding the following fix 
https://github.com/cssinjs/repl/pull/5

**Comment by pranshuchittora** [comment]
Till v10.0.0 it was working perfectly. In `v10.0.1` the issue of dynamicStyles is occurring.

The line -> https://github.com/cssinjs/jss/blob/d9cca31ad2d058aa0108fa6ec48fe872b0e7e886/packages/jss/src/DomRenderer.js#L406

Is trying to concert the dynamic style to string, which is unable to.

![Screenshot from 2020-05-07 01-34-12](https://user-images.githubusercontent.com/32242596/81223167-35cab480-9003-11ea-8a7f-e53e6a3478a2.png)

This flow is encountered when the `link` is true.
IMO the issue is in `jss` not `react-jss`.

P.S. `update` method on `rule` is not available.

**Comment by kof** [comment]
That line was changed last time before 10.0.0

**Comment by pranshuchittora** [comment]
Hi @kof ,
So I ran `git bisect` between the `v10.0.0` & `v10.0.1`. And found that the commit which introduced that bug was [9f7924eb79bf1e78816db7aaa4cf324471](https://github.com/cssinjs/jss/commit/9f7924eb79bf1e78816db7aaa4cf324471).

![Screenshot from 2020-05-09 14-25-42](https://user-images.githubusercontent.com/32242596/81469474-a0960e80-9202-11ea-8a79-b4c0983f3d28.png)

Reverting the changes in `packages/react-jss/src/createUseStyles.js` introduced in the following commit fixes the bug.

**Comment by kof** [comment]
Can you create a test? And optionally a fix?

**Comment by pranshuchittora** [comment]
@kof kindly check #1343  for the fixes :)

**Comment by jmikrut** [comment]
This is still an issue, correct? I am on `10.6.0` and can reproduce.

**Comment by kof** [comment]
https://github.com/cssinjs/jss/pull/1343#issuecomment-808819503

**Comment by mjeanroy** [comment]
Hi @Kof,

I tried to debug this, and I would like to share with you what I found.
With my coworker @pybuche, we tried to understand what was the issue here, and he found something interesting: if we [wrap the call to `manageSheet`](https://github.com/cssinjs/jss/blob/38075fa0085922872045070941ebf2686038db79/packages/react-jss/src/createUseStyles.js#L54) in a `setTimeout`, the issue disappears!
We tried to understand why, and here is my explanation:

1. When the first component renders, everything works fine. In fact, the sheet is correct, and all the rules are deployed using `manageSheet`, the sheet is then marked as attached.

2. When the second component renders, the sheet is already attached, but all its dynamic rule are updated.

If we wrap `manageSheet` in a `setTimeout`, then, when the second component renders, the sheet is not attached yet to the DOM, so it works because all the rules are attached & deployed properly.

The issue really occurs when updating a dynamic rule, in fact, if we do:

```js
if (newSheet) {
  // Detach the sheet so all rules will be inserted
  unmanageSheet({
    index,
    context,
    sheet: newSheet,
    theme
  });

  manageSheet({
    index,
    context,
    sheet: newSheet,
    theme
  });
}
```

=> It also fixes the issue.

I tried to go deeper to understand why updating dynamic rules did not works as exepected, and I think the issue is in `jss-plugin-nested`, especially [here](https://github.com/mjeanroy/jss/blob/38075fa0085922872045070941ebf2686038db79/packages/jss-plugin-nested/src/index.js#L101):

```js
} else if (isNestedConditional) {
  // Place conditional right after the parent rule to ensure right ordering.
  container
    .addRule(prop, {}, options)
    // Flow expects more options but they aren't required
    // And flow doesn't know this will always be a StyleRule which has the addRule method
    // $FlowFixMe[incompatible-use]
    // $FlowFixMe[prop-missing]
    .addRule(styleRule.key, style[prop], {selector: styleRule.selector})
}
```

The line `container.addRule` insert rule, but in fact, the rule is not yet up to date, since the child rule is added the line after (to summarize, `insertRule` happens too early I think).

I tried to fix it by splitting the rule addition and the insertion, something like this:

```js
const addedRule = container.createRule(prop, {}, options);
addedRule.addRule(styleRule.key, style[prop], {selector: styleRule.selector})
container.deployRule(addedRule);
```

You can find the commit on my fork [here](https://github.com/mjeanroy/jss/commit/46493add8ef5613e8603aca50a67f388c1d256ee). Note that it is just a POC, I did not try to make it "super clean", but if you think the solution seems OK, feel free to give your opinion and I can work on a PR :)

In the meantime, I would like to suggest a workaround until the issue is fixed: adding a plugin that detach and re-attach the sheet, such as:

```js
jss.use({
  onUpdate(data: any, rule, sheet, options) {
    if (sheet) {
      sheet.detach().attach();
    }
  },
});
```

This workaround can definitely be improved, that is just the basic idea :)

_Also, thanks for your amazing library, that's an impressive work, and I really love using it_

**Comment by hkar19** [comment]
definitely @mjeanroy suggestion to use:
```js
jss.use({
  onUpdate(data: any, rule, sheet, options) {
    if (sheet) {
      sheet.detach().attach();
    }
  },
});
``` 

is very nice.

I just want to add, if you are using react-jss, remember that using [custom setup](https://cssinjs.org/react-jss/?v=v10.7.1#custom-setup) i.e. utilizing `jss.use`, you need to do things manually such as:

1. dont forget to call `preset` from `jss-preset-default` in your `createJss`, such as:
```ts
import { JssProvider, ThemeProvider } from "react-jss";
import preset from "jss-preset-default";

// i made my array here, and spread it in my jss.use
import jssPlugins from "./jssPlugins";

// dont forget this!
const jss = createJss(preset());
jss.use(
  {
    onUpdate(data: object, rule: Rule, sheet: StyleSheet) {
      if (sheet) {
        sheet.detach().attach();
      }
    },
  },
  ...jssPlugins
);
```

2. in jss plugins, i do:
```ts
import a from "jss-plugin-camel-case";
import b from "jss-plugin-compose";
import c from "jss-plugin-default-unit";
import d from "jss-plugin-expand";
import e from "jss-plugin-extend";
import f from "jss-plugin-global";
import g from "jss-plugin-nested";
import h from "jss-plugin-props-sort";
import i from "jss-plugin-rule-value-function";
import j from "jss-plugin-rule-value-observable";
import k from "jss-plugin-template";
import l from "jss-plugin-vendor-prefixer";

export default [a(), b(), c(), d(), e(), f(), g(), h(), i(), j(), k(), l()];

```

this workaround is tedious, but for now this is all we got

**Comment by j4mesjung** [comment]
Hello, any updates on a fix for this?  Experiencing this issue on `10.9.0`

**Comment by TchernyavskyDaniil** [comment]
Sorry if I offended or don't understand your processes, I'm speaking as a regular consumer of your library

This bug is from version 10.0.1. Shouldn't it be fixed by the current library team? Part of my team is giving up on jss, as well as some people I know because of this problem. I honestly don't quite understand the point of preparing a move to 18 React if there are such critical bugs

**Comment by kof** [comment]
I am merging good PRs that make progress, if a PR ads support for react 18 and has tests etc - cool.

if you want to get something fixed, please fix it, add a test - I will merge it too.
